modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgnitePlanner.java

```diff
@@ -231,17 +238,91 @@ public RelNode convert(SqlNode sql) {
     }
 
     /**
-     * Validates a SQL statement.
+     * Validates a SQL statement and tries to build aliases for appropriate column names.
      *
      * @param sqlNode Root node of the SQL parse tree.
      * @return Validated node, its validated type and type's origins.
      */
-    public ValidationResult validateAndGetTypeMetadata(SqlNode sqlNode) {
+    ValidationResult validateAndGetTypeMetadata(SqlNode sqlNode) {
+        List<SqlNode> selectItems = null;
+        List<SqlNode> selectItemsNoStar = null;
+        SqlNode sqlNode0 = sqlNode instanceof SqlOrderBy ? ((SqlOrderBy) sqlNode).query : sqlNode;
+
+        boolean starFound = false;
+
+        if (sqlNode0 instanceof SqlSelect) {
+            selectItems = SqlNonNullableAccessors.getSelectList((SqlSelect) sqlNode0);
+            selectItemsNoStar = new ArrayList<>(selectItems.size());
+
+            for (SqlNode node : selectItems) {
+                if (node instanceof SqlIdentifier) {
+                    SqlIdentifier id = (SqlIdentifier) node;
+                    if (id.isStar()) {
+                        starFound = true;
+                    }
+                } else {
+                    if (node instanceof SqlBasicCall) {
+                        SqlBasicCall node0 = (SqlBasicCall) node;
+                        if (!identAsIdent(node0)) {
+                            selectItemsNoStar.add(node);
+                        }
+                    } else {
+                        selectItemsNoStar.add(node);
+                    }
+                }
+            }
+        }
+
         SqlNode validatedNode = validator().validate(sqlNode);
         RelDataType type = validator().getValidatedNodeType(validatedNode);
         List<List<String>> origins = validator().getFieldOrigins(validatedNode);
 
-        return new ValidationResult(validatedNode, type, origins);
+        List<String> derived = null;
+        if (validatedNode instanceof SqlSelect && selectItems != null && !selectItemsNoStar.isEmpty()) {
+            derived = new ArrayList<>(selectItems.size());
+
+            if (starFound) {
+                SqlNodeList expandedItems = ((SqlSelect) validatedNode).getSelectList();
+
+                int resolved = 0;
+                for (SqlNode node : expandedItems) {
+                    if (node instanceof SqlIdentifier) {
```

Comment:
Personally I would prefer to combine the following two conditions into one, something like
```
if (node instanceof SqlIdentifier || identAsIdent(node)) {
```
but it's up to you.